From 2d40a32e35981a77ac4a6d2a289957e8e5216742 Mon Sep 17 00:00:00 2001
From: Dong Chenchen <superdcc97@163.com>
Date: Wed, 13 May 2026 00:25:50 +0800
Subject: [PATCH 07/13] mm: keep i_mmap contention counter in address_space

Dynamic i_mmap sharding needs to count failed trylock attempts before
sharded state exists.  Keeping that counter inside struct i_mmap_shards
makes the first-contention path depend on allocating the structure
it is trying to justify.

Move the trylock failure counter into struct address_space and
initialize it with the rest of the mapping state.  The sharded
allocation helper now only initializes state that belongs to an
allocated shard set.

Signed-off-by: Dong Chenchen <superdcc97@163.com>
---
 fs/inode.c         | 1 +
 include/linux/fs.h | 3 ++-
 mm/vma.c           | 1 -
 3 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/fs/inode.c b/fs/inode.c
index 791ce61c2f64..4acf939756ef 100644
--- a/fs/inode.c
+++ b/fs/inode.c
@@ -279,6 +279,7 @@ int inode_init_always_gfp(struct super_block *sb, struct inode *inode, gfp_t gfp
 	mapping->flags = 0;
 	mapping->wb_err = 0;
 	atomic_set(&mapping->i_mmap_writable, 0);
+	atomic_set(&mapping->i_mmap_trylock_fails, 0);
 #ifdef CONFIG_READ_ONLY_THP_FOR_FS
 	atomic_set(&mapping->nr_thps, 0);
 #endif
diff --git a/include/linux/fs.h b/include/linux/fs.h
index 5fb6060dec15..b6988be2b876 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -461,7 +461,6 @@ struct i_mmap_shard {
 struct i_mmap_shards {
 	unsigned int		nr_shards;
 	atomic_t		nr_vmas;
-	atomic_t		trylock_fails;
 	struct i_mmap_shard	shard[];
 };
 
@@ -481,6 +480,7 @@ void i_mmap_shards_free(struct i_mmap_shards *shards);
  * @nr_thps: Number of THPs in the pagecache (non-shmem only).
  * @i_mmap: Tree of private and shared mappings.
  * @i_mmap_shards: Optional distributed i_mmap trees and locks.
+ * @i_mmap_trylock_fails: Contention counter for dynamic i_mmap sharding.
  * @i_mmap_rwsem: Protects @i_mmap and @i_mmap_writable.
  * @nrpages: Number of page entries, protected by the i_pages lock.
  * @writeback_index: Writeback starts here.
@@ -501,6 +501,7 @@ struct address_space {
 #endif
 	struct rb_root_cached	i_mmap;
 	struct i_mmap_shards	*i_mmap_shards;
+	atomic_t		i_mmap_trylock_fails;
 	unsigned long		nrpages;
 	pgoff_t			writeback_index;
 	const struct address_space_operations *a_ops;
diff --git a/mm/vma.c b/mm/vma.c
index efee0d0c4805..c8249e99ae75 100644
--- a/mm/vma.c
+++ b/mm/vma.c
@@ -18,7 +18,6 @@ struct i_mmap_shards *i_mmap_shards_alloc(gfp_t gfp)
 
 	shards->nr_shards = I_MMAP_NR_SHARDS;
 	atomic_set(&shards->nr_vmas, 0);
-	atomic_set(&shards->trylock_fails, 0);
 
 	for (i = 0; i < I_MMAP_NR_SHARDS; i++) {
 		shards->shard[i].tree = RB_ROOT_CACHED;
-- 
2.43.0

